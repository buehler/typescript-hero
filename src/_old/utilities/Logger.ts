import { ExtensionConfig } from '../ExtensionConfig';
import { injectable } from 'inversify';
import * as util from 'util';
import { ExtensionContext, OutputChannel, window } from 'vscode';

/**
 * Factory type for IoC.
 * 
 * @typedef LoggerFactory
 */
export type LoggerFactory = (prefix?: string) => Logger;

/**
 * Enum for the log level. Determines if a message is logged into the output.
 * 
 * @export
 * @enum {number}
 */
export const enum LogLevel {
    Nothing,
    Errors,
    Warnings,
    All
}

/**
 * Central logger instance of the extension.
 * 
 * @export
 * @class Logger
 */
@injectable()
export class Logger {
    private static channel: OutputChannel;

    constructor(context: ExtensionContext, private config: ExtensionConfig, private prefix?: string) {
        if (!Logger.channel) {
            Logger.channel = window.createOutputChannel('Typescript Hero Extension');
            context.subscriptions.push(Logger.channel);
        }
    }

    /**
     * Logs an error message. Provided data is logged out after the message.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    public error(message: string, data?: any): void {
        this.log(LogLevel.Errors, `ERROR\t ${message}`, data);
    }

    /**
     * Logs a warning message. Provided data is logged out after the message.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    public warning(message: string, data?: any): void {
        this.log(LogLevel.Warnings, `WARNING\t ${message}`, data);
    }

    /**
     * Logs an info message. Provided data is logged out after the message.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    public info(message: string, data?: any): void {
        this.log(LogLevel.All, `INFO\t ${message}`, data);
    }

    /**
     * Internal method to actually do the logging. Checks if the output should be done and logs
     * the data into the output channel and the console (if debugging).
     * 
     * @private
     * @param {LogLevel} level
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    private log(level: LogLevel, message: string, data?: any): void {
        if (this.config.logLevel >= level) {
            // tslint:disable-next-line
            console.log(`${this.prefix ? `${this.prefix}: ` : ''}${message}`, data);
            Logger.channel.appendLine(`${this.prefix ? `${this.prefix}: ` : ''}${message}`);
            if (data) {
                Logger.channel.appendLine(`\tData:\t${util.inspect(data, {})}`);
            }
        }
    }
}
