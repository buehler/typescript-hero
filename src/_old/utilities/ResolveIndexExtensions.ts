import { DeclarationInfo, ResolveIndex } from '../caches/ResolveIndex';
import { DefaultDeclaration } from '../models/TsDeclaration';
import {
    TsDefaultImport,
    TsExternalModuleImport,
    TsImport,
    TsNamedImport,
    TsNamespaceImport
} from '../models/TsImport';
import { ImportLocation } from '../models/TsImportOptions';
import { join, normalize, parse, relative } from 'path';
import { Position, TextEditor, workspace } from 'vscode';

/**
 * Calculate the position, where a new import should be inserted.
 * Does respect the "use strict" string as first line of a document.
 * 
 * @export
 * @param {ImportLocation} location
 * @param {TextEditor} editor
 * @returns {Position}
 */
export function getImportInsertPosition(location: ImportLocation, editor: TextEditor): Position {
    if (!editor) {
        return new Position(0, 0);
    }
    if (location === ImportLocation.TopOfFile) {
        return editor.document.lineAt(0).text.match(/use strict/) ? new Position(1, 0) : new Position(0, 0);
    }
    return new Position(editor.selection.active.line, 0);
}
