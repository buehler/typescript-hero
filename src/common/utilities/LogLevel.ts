/**
 * Enum for the log level. Determines if a message is logged into the output.
 * 
 * @export
 * @enum {number}
 */
export const enum LogLevel {
    Nothing,
    Errors,
    Warnings,
    All,
}
