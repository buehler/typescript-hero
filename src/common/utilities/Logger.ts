/**
 * Factory type for IoC.
 * 
 * @typedef LoggerFactory
 */
export type LoggerFactory = (prefix?: string) => Logger;

/**
 * Interface for logging instances, implement the basic methods needed for logging.
 * If a prefix is provided, it should use it before all messages.
 * 
 * @export
 * @interface Logger
 */
export interface Logger {
    /**
     * The prefix used for the messages.
     * 
     * @type {string}
     * @memberOf Logger
     */
    readonly prefix?: string;

    /**
     * Logs a message marked as error. If additional data is provided, it's logged as well.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    error(message: string, data?: any): void;

    /**
     * Logs a message marked as warning. If additional data is provided, it's logged as well.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    warning(message: string, data?: any): void;

    /**
     * Logs a message marked as info. If additional data is provided, it's logged as well.
     * 
     * @param {string} message
     * @param {*} [data]
     * 
     * @memberOf Logger
     */
    info(message: string, data?: any): void;
}
