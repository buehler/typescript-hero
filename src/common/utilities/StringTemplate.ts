/**
 * Creates a template from an expression string. The template can then be used to infuse stuff into the template.
 * 
 * @export
 * @param {string[]} strings
 * @param {...number[]} keys
 * @returns {(...values: any[]) => string}
 */
export function stringTemplate(strings: TemplateStringsArray, ...keys: number[]): (...values: any[]) => string {
    return (...values: any[]) => {
        const result = [strings[0]];

        keys.forEach((key, idx) => {
            result.push(values[key], strings[idx + 1]);
        });

        return result.join('');
    };
}
