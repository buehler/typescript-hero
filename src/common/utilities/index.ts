export * from './LogLevel';
export * from './Logger';
export * from './Clonable';
