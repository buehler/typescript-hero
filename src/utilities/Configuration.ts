import * as vscode from 'vscode';

const sectionKey = 'typescriptHero';

export class Configuration {
    public static get pathStringDelimiter(): string {
        return vscode.workspace.getConfiguration(sectionKey).get<string>('pathStringDelimiter');
    }

    public static get includeNodeModules(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('includeNodeModules');
    }

    public static get includeTypings(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('includeTypings');
    }

    public static get refreshOnSave(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('refreshOnSave');
    }

    // public static get organizeOnSave(): boolean {
    //     return vscode.workspace.getConfiguration(sectionKey).get<boolean>('organizeOnSave');
    // }

    public static get debug(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('debug');
    }
}
