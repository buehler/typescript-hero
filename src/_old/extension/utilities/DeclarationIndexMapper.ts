import { inject, injectable, postConstruct } from 'inversify';
import { DeclarationIndex, FileChanges, TypescriptParser } from 'typescript-parser';
import {
  Event,
  EventEmitter,
  ExtensionContext,
  FileSystemWatcher,
  RelativePattern,
  Uri,
  workspace,
  WorkspaceFolder,
  WorkspaceFoldersChangeEvent,
} from 'vscode';

import { ConfigFactory } from '../../common/factories';
import { findFiles } from '../../common/helpers';
import { iocSymbols } from '../../extension/IoCSymbols';
import { Logger } from './winstonLogger';

interface WorkspaceIndex {
  index: DeclarationIndex;
  folder: WorkspaceFolder;
  watcher: FileSystemWatcher;
}

// TODO move did change configuration to all indices
// TODO: update index on change of configs

/**
 * Mapping class that manages the different indices for the workspaces.
 *
 * @export
 * @class DeclarationIndexMapper
 */
@injectable()
export class DeclarationIndexMapper {
  public readonly onStartIndexing: Event<WorkspaceIndex>;
  public readonly onFinishIndexing: Event<WorkspaceIndex>;
  public readonly onIndexingError: Event<{ index: WorkspaceIndex, error: Error }>;

  private _onStartIndexing: EventEmitter<WorkspaceIndex>;
  private _onFinishIndexing: EventEmitter<WorkspaceIndex>;
  private _onIndexingError: EventEmitter<{ index: WorkspaceIndex, error: Error }>;

  private indizes: { [uri: string]: WorkspaceIndex } = {};

  constructor(
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.typescriptParser) private parser: TypescriptParser,
    @inject(iocSymbols.configuration) private config: ConfigFactory,
  ) {
    this._onFinishIndexing = new EventEmitter();
    this._onStartIndexing = new EventEmitter();
    this._onIndexingError = new EventEmitter();

    this.onFinishIndexing = this._onFinishIndexing.event;
    this.onStartIndexing = this._onStartIndexing.event;
    this.onIndexingError = this._onIndexingError.event;

    this.context.subscriptions.push(this._onFinishIndexing);
    this.context.subscriptions.push(this._onIndexingError);
    this.context.subscriptions.push(this._onStartIndexing);
  }

  /**
   * Init method that runs after the DI construction of the object.
   *
   * @memberof DeclarationIndexMapper
   */
  @postConstruct()
  public initialize(): void {
    this.context.subscriptions.push(workspace.onDidChangeWorkspaceFolders(e => this.workspaceChanged(e)));
    this.logger.info(
      '[%s] initializing index mapper for %d workspaces',
      DeclarationIndexMapper.name,
      (workspace.workspaceFolders || []).length,
    );

    for (const folder of (workspace.workspaceFolders || []).filter(workspace => workspace.uri.scheme === 'file')) {
      this.initializeIndex(folder);
    }
    this.logger.info('[%s] initialized', DeclarationIndexMapper.name);
  }

  /**
   * Method to rebuild all indices in the system.
   *
   * @memberof DeclarationIndexMapper
   */
  public rebuildAll(): void {
    this.logger.info(
      '[%s] rebuilding all indices',
      DeclarationIndexMapper.name,
    );
    for (const index of Object.values(this.indizes)) {
      index.watcher.dispose();
      index.index.reset();
    }
    this.indizes = {};
    for (const folder of (workspace.workspaceFolders || []).filter(workspace => workspace.uri.scheme === 'file')) {
      this.initializeIndex(folder);
    }
  }

  /**
   * Returns the index (or undefined) for a given file URI.
   *
   * @param {Uri} fileUri
   * @returns {(DeclarationIndex | undefined)}
   * @memberof DeclarationIndexMapper
   */
  public getIndexForFile(fileUri: Uri): DeclarationIndex | undefined {
    const workspaceFolder = workspace.getWorkspaceFolder(fileUri);
    if (!workspaceFolder || !this.indizes[workspaceFolder.uri.fsPath]) {
      this.logger.debug(
        '[%s] did not find index for file',
        DeclarationIndexMapper.name,
        { file: fileUri.fsPath },
      );
      return;
    }

    return this.indizes[workspaceFolder.uri.fsPath].index;
  }

  /**
   * Eventhandler that is called when the workspaces changed (i.e. some where added or removed).
   *
   * @private
   * @param {WorkspaceFoldersChangeEvent} event
   * @memberof DeclarationIndexMapper
   */
  private workspaceChanged(event: WorkspaceFoldersChangeEvent): void {
    this.logger.info(
      '[%s] workspaces changed, adjusting indices',
      DeclarationIndexMapper.name,
    );
    for (const add of event.added) {
      this.logger.debug(
        '[%s] add workspace for "%s"',
        DeclarationIndexMapper.name,
        add.uri.fsPath,
      );
      if (this.indizes[add.uri.fsPath]) {
        this.logger.warn(
          '[%s] workspace index "%s" already exists, skipping',
          DeclarationIndexMapper.name,
          add.uri.fsPath,
        );
        continue;
      }
      this.initializeIndex(add);
    }

    for (const remove of event.removed) {
      this.logger.debug(
        '[%s] remove workspace for "%s"',
        DeclarationIndexMapper.name,
        remove.uri.fsPath,
      );
      this.indizes[remove.uri.fsPath].index.reset();
      this.indizes[remove.uri.fsPath].watcher.dispose();
      delete this.indizes[remove.uri.fsPath];
    }
  }

  /**
   * Helper method to initialize an index.
   *
   * @private
   * @param {WorkspaceFolder} folder
   * @returns {Promise<void>}
   * @memberof DeclarationIndexMapper
   */
  private async initializeIndex(folder: WorkspaceFolder): Promise<void> {
    const profiler = this.logger.startTimer();
    this.logger.debug(
      '[%s] initialize index for "%s"',
      DeclarationIndexMapper.name,
      folder.uri.fsPath,
    );
    const index = new DeclarationIndex(this.parser, folder.uri.fsPath);
    const config = this.config(folder.uri);
    const files = await findFiles(config, folder);
    const watcher = workspace.createFileSystemWatcher(
      new RelativePattern(
        folder,
        `{${config.resolver.resolverModeFileGlobs.join(',')},**/package.json,**/typings.json}`,
      ),
    );
    const workspaceIndex = {
      index,
      folder,
      watcher,
    };

    this._onStartIndexing.fire(workspaceIndex);

    let timeout: NodeJS.Timer | undefined;
    let events: FileChanges | undefined;

    const fileWatcherEvent = (event: string, uri: Uri) => {
      if (timeout) {
        clearTimeout(timeout);
      }
      if (!events) {
        events = {
          created: [],
          updated: [],
          deleted: [],
        };
      }
      events[event].push(uri.fsPath);

      timeout = setTimeout(
        async () => {
          if (events) {
            const profiler = this.logger.startTimer();
            this.logger.debug(
              '[%s] rebuilding index for index "%s"',
              DeclarationIndexMapper.name,
              folder.uri.fsPath,
            );
            await index.reindexForChanges(events);
            profiler.done({
              message: `[${DeclarationIndexMapper.name}] rebuilt index for workspace "${folder.name}"`,
            });
            events = undefined;
          }
        },
        500,
      );
    };

    watcher.onDidCreate(uri => fileWatcherEvent('created', uri));
    watcher.onDidChange(uri => fileWatcherEvent('updated', uri));
    watcher.onDidDelete(uri => fileWatcherEvent('deleted', uri));

    try {
      await index.buildIndex(files);
      this.indizes[folder.uri.fsPath] = workspaceIndex;
      this._onFinishIndexing.fire(workspaceIndex);
      profiler.done({
        message: `[${DeclarationIndexMapper.name}] built index for workspace "${folder.name}"`,
      });
    } catch (error) {
      this.logger.error(
        '[%s] could not build index for workspace "%s", error: %s',
        DeclarationIndexMapper.name,
        folder.uri.fsPath,
        error,
      );
      this._onIndexingError.fire({ error, index: workspaceIndex });
    }
  }
}
