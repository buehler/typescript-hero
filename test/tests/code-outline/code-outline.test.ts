import { join } from 'path';
import { TypescriptParser } from 'typescript-parser';
import { commands, ExtensionContext, window, workspace } from 'vscode';

import CodeOutline from '../../../src/code-outline';
import Configuration from '../../../src/configuration';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';
import { Logger } from '../../../src/utilities/logger';
import { expect } from '../setup';

describe('CodeOutline', () => {
  const rootPath = workspace.workspaceFolders![0].uri.fsPath;
  const files = {
    empty: join(rootPath, 'code-outline', 'empty.ts'),
    nonParseable: join(rootPath, 'code-outline', 'not-parseable.txt'),
    codeTs: join(rootPath, 'code-outline', 'code.ts'),
    codeJs: join(rootPath, 'code-outline', 'code.js'),
  };

  let extension: CodeOutline;

  before(async () => {
    const document = await workspace.openTextDocument(files.empty);
    await window.showTextDocument(document);

    const context = ioc.get<ExtensionContext>(iocSymbols.extensionContext);
    const logger = ioc.get<Logger>(iocSymbols.logger);
    const config = ioc.get<Configuration>(iocSymbols.configuration);
    const parser = ioc.get<TypescriptParser>(iocSymbols.parser);

    extension = new CodeOutline(context, logger, config, parser);
  });

  after(async () => {
    await commands.executeCommand('workbench.action.closeAllEditors');
  });

  it('should return an empty array if no document is open', async () => {
    await commands.executeCommand('workbench.action.closeAllEditors');
    expect(await extension.getChildren()).to.matchSnapshot();
  });

  it('should return disabled structure item when it is disabled', async () => {
    const document = await workspace.openTextDocument(files.codeTs);
    await window.showTextDocument(document);
    const config = workspace.getConfiguration('typescriptHero');
    const value = config.get('codeOutline.enabled', true);

    try {
      await config.update('codeOutline.enabled', false);
      expect(
        (await extension.getChildren() || []).map(c => ({ label: c.label, ctor: c.constructor.name })),
      ).to.matchSnapshot();
    } finally {
      await config.update('codeOutline.enabled', value);
    }
  });

  it('should return not parseable on non ts / js file', async () => {
    const document = await workspace.openTextDocument(files.nonParseable);
    await window.showTextDocument(document);

    expect(
      (await extension.getChildren() || []).map(c => ({ label: c.label, ctor: c.constructor.name })),
    ).to.matchSnapshot();
  });

  it('should return a code structure for a valid ts file', async () => {
    const document = await workspace.openTextDocument(files.codeTs);
    await window.showTextDocument(document);
    (extension as any).documentCache = undefined;
    expect(
      (await extension.getChildren() || []).map(c => ({ label: c.label, ctor: c.constructor.name })),
    ).to.matchSnapshot();
  });

  it('should return a code structure for a valid js file', async () => {
    const document = await workspace.openTextDocument(files.codeJs);
    await window.showTextDocument(document);
    (extension as any).documentCache = undefined;
    expect(
      (await extension.getChildren() || []).map(c => ({ label: c.label, ctor: c.constructor.name })),
    ).to.matchSnapshot();
  });

});
