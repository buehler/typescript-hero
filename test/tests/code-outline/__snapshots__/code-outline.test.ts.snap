// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeOutline should return a code structure for a valid js file 1`] = `
Array [
  Object {
    "ctor": "ImportsStructureTreeItem",
    "label": "Imports",
  },
  Object {
    "ctor": "DeclarationStructureTreeItem",
    "label": "Yay",
  },
]
`;

exports[`CodeOutline should return a code structure for a valid ts file 1`] = `
Array [
  Object {
    "ctor": "ImportsStructureTreeItem",
    "label": "Imports",
  },
  Object {
    "ctor": "DeclarationStructureTreeItem",
    "label": "Yay",
  },
  Object {
    "ctor": "DeclarationStructureTreeItem",
    "label": "ArrList<T>",
  },
]
`;

exports[`CodeOutline should return an empty array if no document is open 1`] = `Array []`;

exports[`CodeOutline should return disabled structure item when it is disabled 1`] = `
Array [
  Object {
    "ctor": "DisabledStructureTreeItem",
    "label": "Feature is disabled.",
  },
]
`;

exports[`CodeOutline should return not parseable on non ts / js file 1`] = `
Array [
  Object {
    "ctor": "NotParseableStructureTreeItem",
    "label": "File not parseable.",
  },
]
`;
